dbt/adapters/base/impl.py

```diff
@@ -1285,46 +1287,92 @@ def calculate_freshness(
         }
         return adapter_response, freshness
 
+    def calculate_freshness_from_metadata_batch(
+        self,
+        sources: List[BaseRelation],
+        macro_resolver: Optional[MacroResolverProtocol] = None,
+    ) -> Tuple[Optional[AdapterResponse], Dict[BaseRelation, FreshnessResponse]]:
+        # Track schema, identifiers of sources for lookup from batch query
+        schema_identifier_to_source = {
+            (
+                source.path.get_lowered_part(ComponentName.Schema),
+                source.path.get_lowered_part(ComponentName.Identifier),
+            ): source
+            for source in sources
+        }
+
+        # Group metadata sources by information schema -- one query per information schema will be necessary
+        information_schema_to_metadata_sources: Dict[InformationSchema, List[BaseRelation]] = {}
+        for source in sources:
+            information_schema = source.information_schema_only()
+            if information_schema not in information_schema_to_metadata_sources:
+                information_schema_to_metadata_sources[information_schema] = [source]
+            else:
+                information_schema_to_metadata_sources[information_schema].append(source)
+
+        freshness_responses: Dict[BaseRelation, FreshnessResponse] = {}
+        for (
+            information_schema,
+            sources_for_information_schema,
+        ) in information_schema_to_metadata_sources.items():
+            result = self.execute_macro(
+                GET_RELATION_LAST_MODIFIED_MACRO_NAME,
+                kwargs={
+                    "information_schema": information_schema,
+                    "relations": sources_for_information_schema,
+                },
+                macro_resolver=macro_resolver,
+            )
+            adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
+
+            for row in table:
```

Comment:
Have you considered putting the contents of this `for` loop in a helper function? There are a few reasons I'd consider it here. This is a nested for loop and there's a decent amount of processing happening within the loop. Separating the logic may make it easier to read. I would also consider doing the relation mapping outside of the loop. For example:

```python
def calculate_freshness_from_metadata_batch(...):

    ...

    raw_freshness_responses: Dict[Tuple[str, str], FreshnessResponse] = {}
    for (
        information_schema,
        sources_for_information_schema,
    ) in information_schema_to_metadata_sources.items():
        result = self.execute_macro(
            GET_RELATION_LAST_MODIFIED_MACRO_NAME,
            kwargs={
                "information_schema": information_schema,
                "relations": sources_for_information_schema,
            },
            macro_resolver=macro_resolver,
        )
        adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
        raw_freshness_responses.update(
            self._parse_freshness_row(row)
            for row in table
        )
    
    # handle the schema/identifier -> BaseRelation here
    
    return adapter_response, freshness_response

...

def _parse_freshness_row(self, row: agate.Row) -> Tuple[Any, FreshnessResponse]:
    try:
        last_modified_val = get_column_value_uncased("last_modified", row)
        snapshotted_at_val = get_column_value_uncased("snapshotted_at", row)
        identifier = get_column_value_uncased("identifier", row)
        schema = get_column_value_uncased("schema", row)
    except Exception:
        raise MacroResultError(GET_RELATION_LAST_MODIFIED_MACRO_NAME, table)
    
    freshness_response = self._create_freshness_response(
        last_modified_val, snapshotted_at_val
    )
    raw_relation = schema.lower(), identifier.lower()
    return raw_relation, freshness_response
```

Reply:
Incorporated this suggestion, thank you!